library/std/src/sys/wasi/alloc.rs

```diff
@@ -7,36 +9,46 @@ unsafe impl GlobalAlloc for System {
     #[inline]
     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
         if layout.align() <= MIN_ALIGN && layout.align() <= layout.size() {
-            libc::malloc(layout.size()) as *mut u8
+            // SAFETY: The safety contract for `malloc` must be upheld by the caller
+            unsafe { libc::malloc(layout.size()) as *mut u8 }
         } else {
-            libc::aligned_alloc(layout.align(), layout.size()) as *mut u8
+            // SAFETY: The safety contract for `aligned_alloc` must be upheld by the caller
```

Comment:
I would prefer that these be worded something like "`aligned_alloc` has the same requirements as `GlobalAlloc::alloc`" if that's true. If not, we should say which requirements are different (perhaps something to do with alignment?)

This is also currently missing a safety comment on the impl itself justifying the `unsafe` there, which we should add.

Reply:
I think you meant if `aligned_alloc` and `malloc` have the same requirements. I don't think they have the same requirement.

from manpages:
> The obsolete function memalign() allocates size bytes and returns a  pointer  to  the  allocated
    memory.  The memory address will be a multiple of alignment, which must be a power of two.

> The  function  aligned_alloc()  is the same as memalign(), except for the added restriction that
    size should be a multiple of alignment.

About the safety for `unsafe impl`, do you have a suggestion of what should I add? As most functions rely on the caller of safety.

Reply:
I am saying that the current safety comments are incorrect -- we cannot arbitrarily place requirements on our caller. The only "special" knowledge implementors of GlobalAlloc::alloc have is that [the passed Layout is non zero sized](https://doc.rust-lang.org/nightly/alloc/alloc/trait.GlobalAlloc.html#safety-1).

In this case, calling `libc::malloc` should be entirely safe, AFAICT -- there are no preconditions that I can see in the man page. The safety comment on the malloc call should say something to that effect.

With regards to aligned_alloc, the "size should be a multiple of alignment" condition does not seem to be checked in this function, and I cannot find any reason that we can assume that is true. Do you have evidence otherwise? If not, we should likely file an issue and leave a FIXME here.

As to the `unsafe impl`, there should be a comment noting that each function in the trait has been checked to meet the requirements listed in the [safety section of the trait docs](https://doc.rust-lang.org/nightly/alloc/alloc/trait.GlobalAlloc.html#safety). Obviously, we should make sure to actually do so :)

Reply:
Ok, so `malloc` call is safe because it's guaranteed that non zero layout size will be passed.

Like you said there is no check here for layout `size` being a multiple of `alignment`, but is there  any check from the caller? If not, then maybe as you said we should file an issue for it.

Reply:
> Ok, so malloc call is safe because it's guaranteed that non zero layout size will be passed.

This does not seem to be a precondition of malloc, AFAICT. malloc call is just always safe.

We cannot assume that there is a check in the caller in unsafe code, unless the method documents such a precondition. GlobalAlloc::alloc does not. Filing an issue would be good.